### PR TITLE
Refactor lm_head losses

### PR DESCRIPTION
**Pull request by oleksost**
# ✨ Description

Refactors loss definition and logging in the`head.py`:
- makes logging more explicit
- implements forward KL
- only log unscaled losses
- defines a separate loss configs for each loss in the lm+_head
- allows logging of losses without training on them

TODO:
- [x] Update tests

Config example:
```
    head:
      lr_scale: 0.0
      losses:
        lm_loss:
          type: cross_entropy
          factor: 1.0
        reverse_kl:
          type: reverse_kl_distillation
          factor: 1.0
        forward_kl:
          type: forward_kl_distillation
          factor: 0.0 # track without logging
```
This will train using `lm_loss` which is a `cross_entropy_lm_loss` as well as `reverse_kl`, both weighted with 1. Will also log `forward_kl`.

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Change A
2. Change B

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/contributing/contributing).
- [ ] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

File: fast_llm/layers/language_model/head.py [hunk 1]
```diff
@@ -405,31 +405,106 @@ def _logits_cross_entropy_forward_backward(
...
 
-        # TODO: Accumulate grads in-place to reduce memory and compute overhead.
-        grad = _add_tensors(dpo_grad, lm_grad, distillation_grad)
+        return loss, output_parallel_linear_backward(grad, context) if self.training else None
 
-        # TODO: Return individual losses?
-        loss = _add_tensors(dpo_loss, lm_loss, distillation_loss)
-        if self.training and losses is not None:
-            if dpo_loss is not None:
+    def _post_process_loss_and_grad(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py]: This contains lots of unnecessary code. The gradients are conceptually ceparated from the losses, and processing them here adds unnecessary complexity to what is essentially call to `_add_tensors`.

File: fast_llm/layers/language_model/head.py [hunk 2]
```diff
@@ -113,6 +113,12 @@ def __init__(
             peft=self._peft,
         )
 
+        self._compute_lm_loss = self.config.language_model_loss_factor > 0.0 or self.config.track_language_model_loss
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py]: This and `_compute_distillation_loss` would be more appropriate in the config, ex. as a `cachedproperety`

File: fast_llm/layers/language_model/config.py [hunk 3]
```diff
@@ -135,44 +135,22 @@ class LanguageModelHeadConfig(LanguageModelHeadBaseConfig):
...
+        default_factory=dict,
+        desc="A dictionary of loss names and their configurations.",
+        hint=FieldHint.core,
+    )
     # TODO: Cleanup
     output_weight: ParameterConfig = Field(
         desc="Configuration for the LM output layer (weight). Ignored for tied embeddings",
         hint=FieldHint.architecture,
     )
-    cross_entropy_implementation: CrossEntropyImpl = Field(
```
[the remark this answers is not in the document]
Comment (oleksost) [reply fast_llm/layers/language_model/config.py]: I tested training with checkpoints created on the main branch in both `distributed` and `apriel2` format. Training starts with no issues.

File: fast_llm/layers/language_model/lm_head_losses.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+# DPO loss for alignment using chosen and rejected spans.
+#
+
+
+def _format_name(name: str) -> str:
+    return name.replace("_", " ")
+
+
+@dataclasses.dataclass
+class Targets:
```
[the remark this answers is not in the document]
Comment (oleksost) [reply fast_llm/layers/language_model/lm_head_losses.py]: Addressed. 

Each target preparation is handled by loss now in `extract_targets_from_global_kwargs`. This is maybe still a bit suboptimal from the separation of concerns perspective, i.e. the losses now have to handle things related to infra (like splitting_op in case of sequence_parallel_logits)

File: fast_llm/layers/language_model/config.py [hunk 5]
```diff
@@ -236,14 +209,12 @@ def layer_class(self) -> "type[LanguageModelHead]":
...
     def _validate(self) -> None:
-        with self._set_implicit_default():
-            if self.language_model_loss_factor is None:
-                if self.distillation_model is None:
-                    self.language_model_loss_factor = 1.0
-                else:
-                    self.language_model_loss_factor = 0.0
+        for loss_config in self.losses.values():
+            if "dist" in loss_config.type:
+                assert self.distillation_model is not None, "Distillation loss requires a distillation model."
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/layers/language_model/config.py]: The `distillation_model` parameter is not needed in the language model head itself, only the losses use it, so it should be moved with the losses that use it along with these checks.

File: fast_llm/layers/language_model/kwargs.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+from fast_llm.layers.block.config import BlockKwargs
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/kwargs.py]: Not sure about moving here, the convention is to leave in `config.py`

File: fast_llm/layers/language_model/head.py [hunk 7]
```diff
@@ -185,52 +184,22 @@ def _forward_backward(
...
+                continue
+            loss_targets = loss_config.extract_targets_from_global_kwargs(
+                kwargs,
+                prediction_distance=self._prediction_distance,
+                prediction_heads=self._prediction_heads,
+                head_config=self._config,
+                sequence_parallel_logits=self._sequence_parallel_logits,
+                group=self._parallel_dim.group,
+            )
+            targets.update({k: v for k, v in loss_targets.items() if v is not None})
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/head.py]: There could be name conflicts here?

File: fast_llm/layers/language_model/lm_head_losses.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        loss_mask: "torch.Tensor | None",
+        grad_output: float | None = None,
+        group: "ProcessGroup" = None,
+        logits_scale_factor: float | None = None,
+        vocab_parallel: bool = False,
+        kwargs: dict | None = None,
+    ) -> "tuple[torch.Tensor, torch.Tensor | None]":
+        pass
+
+    def get_loss_def(self, name: str, count: int = 1, prediction_distance: int | None = None) -> LossDef:
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/lm_head_losses.py]: `get_loss_definitions`

File: fast_llm/layers/language_model/lm_head_losses.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+    weight: float = Field(
+        default=1.0,
+        hint=FieldHint.core,
+        desc="Weight for this loss in the total loss computation.",
+        valid=check_field(Assert.geq, 0.0),
+    )
+
+    @abc.abstractmethod
+    def compute_loss(
```
Comment (jlamypoirier) [inline fast_llm/layers/language_model/lm_head_losses.py]: `get_loss`

File: tests/test_config.py [hunk 10]
```diff
@@ -296,3 +299,7 @@ def test_distributed_global_ranks(bdp: int, sdp: int, tp: int, pp: int, pipeline
         Assert.eq(len({global_rank for global_ranks in global_ranks_set for global_rank in global_ranks}), world_size)
 
     Assert.eq(len(rank_breakdowns), world_size)
+
+
+if __name__ == "__main__":
```
Comment (jlamypoirier) [inline tests/test_config.py]: Please remove

File: fast_llm/layers/language_model/kwargs.py (hunk 6 above)
Comment (jlamypoirier) [reply fast_llm/layers/language_model/kwargs.py]: How does this lead to circular imports? Was this because of `lm_head_losses.py`?

File: fast_llm/functional/cross_entropy.py [hunk 11]
```diff
@@ -85,6 +85,7 @@ def _fused_cross_entropy_forward_backward(
     target_format: TargetFormat,
     group: ProcessGroup | None = None,
     teacher_softmax_temperature: float = 1.0,
+    return_target_entropy: bool = False,
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/functional/cross_entropy.py]: KL and cross-entropy are basically the same thing, and this is a private method anyway so the name is not that important. I'm more worried about the inconsistent return type and extra complexity this is creating.

File: fast_llm/layers/language_model/kwargs.py (hunk 6 above)
Comment (oleksost) [reply fast_llm/layers/language_model/kwargs.py]: yes, should not be a problem now since `lm_head_losses` was removed

**Comment by jlamypoirier** [comment]
#449
